src/http_server.rs

```diff
@@ -0,0 +1,599 @@
+    warp::any().map(move || bucket.clone())
+}
+
+fn with_s3_client(
+    s3_client: S3Client,
+) -> impl Filter<Extract = (S3Client,), Error = Infallible> + Clone {
+    warp::any().map(move || s3_client.clone())
+}
+
+pub fn esthri_filter(
+    s3_client: S3Client,
+    bucket: &str,
+) -> impl Filter<Extract = (http::Response<Body>,), Error = warp::Rejection> + Clone {
+    warp::path::full()
+        .and(with_s3_client(s3_client))
+        .and(with_bucket(bucket.to_owned()))
+        .and(warp::query::<Params>())
+        .and(warp::header::optional::<String>("if-none-match"))
+        .and_then(download)
+}
+
+pub async fn run(
+    s3_client: S3Client,
+    bucket: &str,
+    address: &SocketAddr,
+) -> Result<(), Infallible> {
+    let routes = esthri_filter(s3_client, bucket).recover(handle_rejection);
+    warp::serve(routes).run(*address).await;
+
+    Ok(())
+}
+
+async fn abort_with_error(
+    archive: Option<&mut Builder<PipeWriter>>,
+    error_tracker: ErrorTrackerArc,
+    err: Report,
+) {
+    let err = {
+        if let Some(archive) = archive {
+            if let Err(err) = archive.finish().await {
+                Report::new(err).wrap_err("closing the archive")
+            } else {
+                err
+            }
+        } else {
+            err
+        }
+    };
+    ErrorTracker::record_error(error_tracker, err);
+}
+
+async fn stream_object_to_archive<T: S3 + Send>(
+    s3: &T,
+    bucket: &str,
+    path: &str,
+    archive: &mut Builder<PipeWriter>,
+    error_tracker: ErrorTrackerArc,
+) -> bool {
+    let obj_info = {
+        match head_object_info(s3, bucket, path).await {
+            Ok(obj_info) => {
+                if let Some(obj_info) = obj_info {
+                    obj_info
+                } else {
+                    abort_with_error(
+                        Some(archive),
+                        error_tracker.clone(),
+                        eyre!("object not found: {}", path),
+                    )
+                    .await;
+                    return !error_tracker.has_error();
+                }
+            }
+            Err(err) => {
+                abort_with_error(
+                    Some(archive),
+                    error_tracker.clone(),
+                    err.wrap_err("s3 head operation failed"),
+                )
+                .await;
+                return !error_tracker.has_error();
+            }
+        }
+    };
+    let stream = {
+        match download_streaming(s3, bucket, path).await {
+            Ok(byte_stream) => byte_stream,
+            Err(err) => {
+                abort_with_error(
+                    Some(archive),
+                    error_tracker.clone(),
+                    err.wrap_err("s3 download failed"),
+                )
+                .await;
+                return !error_tracker.has_error();
+            }
+        }
+    };
+    let mut header = Header::new_gnu();
+    header.set_mode(0o0644);
```

Comment:
Maybe make this a constant?  This is for filetype writing permissions, right?  `0o0644, rw-r--r-- for normal files`

Reply:
Good suggestion, I need to add the last modified date into the archive too, otherwise all the files will show up with a modified date from 1969

Reply:
fixed